### PR TITLE
ClimberSS

### PR DESCRIPTION
**Pull request by laviRZ**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package frc.robot.subsystems.climber;
+
+import edu.wpi.first.math.MathUtil;
+import frc.robot.constants.RobotConstants.ClimberConstants;
+import frc.robot.subsystems.OverridableSubsystem;
+import frc.robot.utilities.pid.PIDFTalonFX;
+
+public class ClimberSS extends OverridableSubsystem {
+    private final PIDFTalonFX rightMotor;
+    private final PIDFTalonFX leftMotor;
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:10]: left before right

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    public ClimberSS() {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:12]: add resetEncoder can call it constructor

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 3]
```diff
@@ -103,6 +104,14 @@
         public static final double RUMBLE_INTERMISSION_TIME = 0.15;
     }
 
+    public static class ClimberConstants {
+        public static final PIDFTalonFX RIGHT_MOTOR = ClimberComponents.RIGHT_MOTOR;
+        public static final PIDFTalonFX LEFT_MOTOR = ClimberComponents.LEFT_MOTOR;
+        public static final DigitalInput RIGHT_DIGITAL_INPUT = ClimberComponents.RIGHT_SWITCH;
+        public static final DigitalInput LEFT_DIGITAL_INPUT = ClimberComponents.LEFT_SWITCH;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:111]: delete

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 4]
```diff
@@ -96,6 +97,13 @@
                 LOCAL_SWERVE_MODULES.rearRightModuleConstants.driveCoefs));
     }
 
+    protected static class ClimberComponents {
+        public static final PIDFTalonFX RIGHT_MOTOR = SwerveComponents.FRONT_RIGHT_ANGLE_MOTOR;
+        public static final PIDFTalonFX LEFT_MOTOR = SwerveComponents.FRONT_LEFT_ANGLE_MOTOR;
+        public static final DigitalInput RIGHT_SWITCH = new DigitalInput(DIO.Climber.RIGHT_SWITCH_CHANNEL);
+        public static final DigitalInput LEFT_SWITCH = new DigitalInput(DIO.Climber.LEFT_SWITCH_CHANNEL);
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:104]: delete

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 5]
```diff
@@ -96,6 +97,13 @@
                 LOCAL_SWERVE_MODULES.rearRightModuleConstants.driveCoefs));
     }
 
+    protected static class ClimberComponents {
+        public static final PIDFTalonFX RIGHT_MOTOR = SwerveComponents.FRONT_RIGHT_ANGLE_MOTOR;
+        public static final PIDFTalonFX LEFT_MOTOR = SwerveComponents.FRONT_LEFT_ANGLE_MOTOR;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:102]: left before right

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 6]
```diff
@@ -40,7 +40,10 @@
     }
 
     protected static class DIO {
-
+        public static class Climber {
+            public static final int LEFT_SWITCH_CHANNEL = 0;
+            public static final int RIGHT_SWITCH_CHANNEL = 1;
+        }
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotMap.java:46]: remove

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 7]
```diff
@@ -77,4 +77,19 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
...
+     * @param degrees
+     * @param radius  of the wheel that move the system or What that
+     * @return the number of rounds that the system has moved
+     */
+    public static double DegreesToMeters(double degrees, double radius) {
+        double circumference = 2 * Math.PI * radius;
+        return degrees * circumference / 360;
+    }
+
+    public static double metersToDegrees(double distance, double radius) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:91]: add documentation

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 8]
```diff
@@ -77,4 +77,19 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
     public static double MagToDegrees(double magTick) {
         return MagToDegrees(magTick, 1);
     }
+
+    /**
+     * @param degrees
+     * @param radius  of the wheel that move the system or What that
+     * @return the number of rounds that the system has moved
+     */
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:85]: fix documentation

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 9]
```diff
@@ -43,13 +43,13 @@ public static double RPMToFalcon(double RPM, double gearRatio) {
     }
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:43]: format this document and also all files you have added/modified

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 10]
```diff
@@ -109,6 +108,12 @@
         public static final double CONTROLLER_DEADBAND = 0.1;
     }
 
+    public static class ClimberConstants {
+        public static final PIDFTalonFX RIGHT_MOTOR = ClimberComponents.RIGHT_MOTOR;
+        public static final PIDFTalonFX LEFT_MOTOR = ClimberComponents.LEFT_MOTOR;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:113]: flip

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 11]
```diff
@@ -40,7 +40,8 @@
     }
 
     protected static class DIO {
-
+        public static class Climber {
+        }
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotMap.java:44]: remove

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        leftMotor = ClimberConstants.LEFT_MOTOR;
+        rightMotor = ClimberConstants.RIGHT_MOTOR;
+    }
+
+    public void resetEncoder() {
+        leftMotor.setSelectedSensorPosition(0);
+        rightMotor.getSelectedSensorPosition(0);
+    }
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:20]: 1. put this below get
    public double getAveragePosition() 
2. call this in constructor

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 13]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
     public static double MagToDegrees(double magTick) {
         return MagToDegrees(magTick, 1);
     }
+
+    /**
+     * @param degrees the degrees of the circle (wheel)
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:82]: "the angle of the wheel or gear in degrees"

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 14]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
...
+     */
+    public static double DegreesToMeters(double degrees, double radius) {
+        double circumference = 2 * Math.PI * radius;
+        return degrees * circumference / 360;
+    }
+
+    /**
+     * @param distance the distance that the object has passed (in meters)
+     * @param radius   the radius of the circle (wheel)
+     * @return the degrees of this distance
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:94]: "the angle of the wheel or gear in degrees"

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 15]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
...
+     * @return the number of rounds that the system has moved
+     */
+    public static double DegreesToMeters(double degrees, double radius) {
+        double circumference = 2 * Math.PI * radius;
+        return degrees * circumference / 360;
+    }
+
+    /**
+     * @param distance the distance that the object has passed (in meters)
+     * @param radius   the radius of the circle (wheel)
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:93]: "the raduis of the wheel or pitch radius of the gear"

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 16]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
     public static double MagToDegrees(double magTick) {
         return MagToDegrees(magTick, 1);
     }
+
+    /**
+     * @param degrees the degrees of the circle (wheel)
+     * @param radius  of the wheel that move the system or What that
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:83]: "the raduis of the wheel or pitch radius of the gear"

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 17]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
...
+     * @return the number of rounds that the system has moved
+     */
+    public static double DegreesToMeters(double degrees, double radius) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:86]: make this camelCase

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 18]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
     public static double MagToDegrees(double magTick) {
         return MagToDegrees(magTick, 1);
     }
+
+    /**
+     * @param degrees the degrees of the circle (wheel)
+     * @param radius  of the wheel that move the system or What that
+     * @return the number of rounds that the system has moved
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:84]: "amount of distance the system has traveled in meters"

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 19]
```diff
@@ -77,4 +77,24 @@ public static double MagToDegrees(double magTicks, double gearRatio) {
...
+     */
+    public static double DegreesToMeters(double degrees, double radius) {
+        double circumference = 2 * Math.PI * radius;
+        return degrees * circumference / 360;
+    }
+
+    /**
+     * @param distance the distance that the object has passed (in meters)
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:92]: "amount of distance the system has traveled in meters"

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotConstants.java (hunk 10 above)
Comment (carmigigangen) [reply src/main/java/frc/robot/constants/RobotConstants.java:113]: שמאלנים

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java (hunk 12 above)
Comment (carmigigangen) [reply src/main/java/frc/robot/subsystems/climber/ClimberSS.java:20]: I did not understand..

Comment (carmigigangen) [reply src/main/java/frc/robot/subsystems/climber/ClimberSS.java]: I did understand..

File: src/main/java/frc/robot/RobotContainer.java [hunk 20]
```diff
@@ -5,7 +5,11 @@
 import edu.wpi.first.wpilibj2.command.InstantCommand;
 import frc.robot.components.TrigonXboxController;
 import frc.robot.constants.RobotConstants.DriverConstants;
+
+import frc.robot.subsystems.climber.ClimberSS;
+
 import frc.robot.subsystems.pitcher.PitcherSS;
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/RobotContainer.java:12]: remove blank lines

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 21]
```diff
@@ -134,6 +134,19 @@
                 CAN.Shooter.RIGHT_MOTOR_ID, MOTOR_CONFIG, ControlMode.Velocity);
     }
 
+    protected static class ClimberComponents {
+        private static final MotorConfig MOTOR_CONFIG = new MotorConfig().
+                brake().
+                inverted(false).
+                withOpenLoopRampRate(0.4).
+                withClosedLoopRampRate(0.4).
+                withPID(RobotConstants.LOCAL_CONSTANTS.localPitcherConstants.pidfCoefs);
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:143]: climber not pitcher

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+    public ClimberSS() {
+        leftMotor = ClimberConstants.LEFT_MOTOR;
+        rightMotor = ClimberConstants.RIGHT_MOTOR;
+        resetEncoder();
+    }
+
+    public double getDesiredPosition() {
+        return leftMotor.getSetpoint();
+    }
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:20]: why?

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+    public void setDesiredPosition(double desiredPosition) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:22]: change name to setSetpoint(position)

Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:22]: add javadoc - position in ticks

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+        leftMotor.setSetpoint(MathUtil.clamp(desiredPosition, 0, ClimberConstants.MAX_POSITION));
+    }
+
+    public double getPositionLeft() {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:26]: leftPosition

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+        return leftMotor.getSelectedSensorPosition();
+    }
+
+    public double getPositionRight() {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:30]: RightPosition

File: src/main/java/frc/robot/utilities/Conversions.java [hunk 26]
```diff
@@ -110,4 +110,24 @@ public static double degreesToMag(double degrees, double gearRatio) {
     public static double kVToTalon(double kV) {
         return kV * 1023;
     }
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/Conversions.java:113]: ctrl + ult + l all files

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/LocalConstants.java [hunk 27]
```diff
@@ -17,6 +17,8 @@
     public LocalPitcherConstants localPitcherConstants;
     @SerializedName("Loader")
     public LocalLoaderConstants localLoaderConstants;
+    @SerializedName("Loader")
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/LocalConstants.java:20]: "Climber" not "Loader"

File: src/main/java/frc/robot/constants/LocalConstants.java [hunk 28]
```diff
@@ -17,6 +17,8 @@
...
+    public LocalClimberConstants localClimberConstants;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/LocalConstants.java:21]: add to constructor

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 29]
```diff
@@ -113,6 +114,12 @@
         public static final PIDFTalonSRX RIGHT_MOTOR = ShooterComponents.RIGHT_MOTOR;
     }
 
+    public static class ClimberConstants {
+        public static final PIDFTalonFX LEFT_MOTOR = ClimberComponents.LEFT_MOTOR;
+        public static final PIDFTalonFX RIGHT_MOTOR = ClimberComponents.RIGHT_MOTOR;
+        public static final int MAX_POSITION = 7000;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:120]: write beside: "// in ticks"

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+    private final PIDFTalonFX rightMotor;
+
+    public ClimberSS() {
+        leftMotor = ClimberConstants.LEFT_MOTOR;
+        rightMotor = ClimberConstants.RIGHT_MOTOR;
+        resetEncoder();
+    }
+
+    /**
+     * @param Position in ticks
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:19]: "@ param position Desired position in ticks"

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public void setSetpoint(double Position) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:21]: 1. parameters should be lower case
2. change position to setpoint

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+        return leftMotor.getSelectedSensorPosition();
+    }
+
+    public double getRightPosition() {
+        return rightMotor.getSelectedSensorPosition();
+    }
+
+    public double getAveragePosition() {
+        return (getLeftPosition() + getRightPosition()) / 2;
+    }
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:35]: add docs:
@ return X position in ticks

File: src/main/java/frc/robot/utilities/Conversions.java [hunk 33]
```diff
@@ -110,4 +110,24 @@ public static double degreesToMag(double degrees, double gearRatio) {
...
+     */
+    public static double degreesToMeters(double degrees, double radius) {
+        double circumference = 2 * Math.PI * radius;
+        return degrees * circumference / 360;
+    }
+
+    /**
+     * @param distance amount of distance the system has traveled in meters
+     * @param radius   radius of the wheel or pitch radius of the gear
+     * @return the angle of the wheel or gear in degrees
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/Conversions.java:127]: Ctrl + alt + l............

**Review by Mboesn — CHANGES_REQUESTED** [review]

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 34]
```diff
@@ -122,7 +122,7 @@
     }
 
     protected static class ShooterComponents {
-        private static MotorConfig MOTOR_CONFIG = new MotorConfig()
+        private static final MotorConfig MOTOR_CONFIG = new MotorConfig()
```
Comment (laviRZ) [inline src/main/java/frc/robot/constants/RobotComponents.java:125]: undo

**Review by laviRZ — COMMENTED** [review]
you have to wait for PIDCMD to implement it

File: src/main/java/frc/robot/constants/SwerveModuleConstants.java [hunk 35]
```diff
@@ -21,4 +21,3 @@ public SwerveModuleConstants(
         this.module = module;
     }
 }
-
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/SwerveModuleConstants.java:24]: why remove line?

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+import frc.robot.utilities.pid.PIDFTalonFX;
+
+public class ClimberSS extends OverridableSubsystem implements PIDSubsystem {
+    private final PIDFTalonFX leftMotor;
+    private final PIDFTalonFX rightMotor;
+
+    public ClimberSS() {
+        leftMotor = ClimberConstants.LEFT_MOTOR;
+        rightMotor = ClimberConstants.RIGHT_MOTOR;
+        resetEncoders();
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:16]: /n

File: src/main/java/frc/robot/subsystems/climber/ClimberSS.java [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * @param setpoint desired position in ticks
+     */
+    @Override
+    public void setSetpoint(double setpoint) {
+        setpoint = MathUtil.clamp(setpoint, 0, ClimberConstants.MAX_POSITION);
+        leftMotor.setSetpoint(setpoint);
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/climber/ClimberSS.java:25]: /n

**Review by laviRZ — COMMENTED** [review]
Approved

**Review by Mboesn — CHANGES_REQUESTED** [review]

**Review by Mboesn — APPROVED** [review]

**Pull request merged**